base/compiler/tfuncs.jl

```diff
@@ -1002,7 +1002,7 @@ swapfield!_tfunc(o, f, v) = (@nospecialize; getfield_tfunc(o, f))
 modifyfield!_tfunc(o, f, op, v, order) = (@nospecialize; modifyfield!_tfunc(o, f, op, v))
 function modifyfield!_tfunc(o, f, op, v)
     @nospecialize
```

Comment:
Unrelated to the PR, do we need to call `widenconst` here? It looks like `_fieldtype_tfunc` might be expecting a Type.
```suggestion
    @nospecialize
    o = widenconst(o)
```
So theoretically that is the cause of this failing to infer currently:
```julia
julia> let x = Ref(0)
       global f() = Core.replacefield!(x, :x, 0, 1)
       end
f (generic function with 1 method)

julia> @code_typed optimize=false f()
CodeInfo(
    @ REPL[24]:2 within `f`
1 ─ %1 = Core.replacefield!::Core.Const(replacefield!)
│   %2 = $(QuoteNode(Base.RefValue{Int64}(0)))
│   %3 = (%1)(%2, :x, 0, 1)::NamedTuple{(:old, :success), Tuple{_A, Bool}} where _A
└──      return %3
) => NamedTuple{(:old, :success), Tuple{_A, Bool}} where _A
```

Reply:
Yes, probably.